src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java

```diff
@@ -1128,6 +1133,123 @@ void verifyJar(String jarName)
         }
     }
 
+    private void crossCheckEntries(String jarName) throws Exception {
+        List<String> locEntries = new ArrayList<>();
+        List<String> cenEntries;
+
+        try (JarFile jarFile = new JarFile(jarName);
+             JarInputStream jis = new JarInputStream(
+                     Files.newInputStream(Path.of(jarName)))) {
+
+            Manifest cenManifest = jarFile.getManifest();
+            Manifest locManifest = jis.getManifest();
+            if (!compareManifest(cenManifest, locManifest)) {
+                return;
+            }
+
+            JarEntry locEntry;
+            while ((locEntry = jis.getNextJarEntry()) != null) {
+                String entryName = locEntry.getName();
+                locEntries.add(entryName);
+
+                JarEntry cenEntry = jarFile.getJarEntry(entryName);
+                if (cenEntry == null) {
+                    crossChkWarnings.add(String.format(rb.getString(
+                            "Entry.missing.in.JarFile.1"), entryName));
+                    continue;
+                }
+
+                readEntry(jis);
+                try (InputStream cenInputStream = jarFile.getInputStream(cenEntry)) {
```

Comment:
What if this returns null? Shouldn't you also emit a warning that the entry was read by JarInputStream but not JarFile? Also, I think `readEntry()` would throw an NPE.

Reply:
Yes, added checking on `cenInputStream` to emit a warning if it is null.